### PR TITLE
feat(Podsights): Onboarding new destination

### PR DESCRIPTION
**Pull request by sauravlal15**
## Description of the change

> Onboarding new destination

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ✅] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

- - -
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-sdk-js/709)

### CODE REVIEW

**Review by MoumitaM — DISMISSED** [review]

File: integrations/Podsights/browser.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+      return;
+    }
+    if (this.enableAliasCall && event.trim().toLowerCase() === "alias") {
+      const externalId =
+        get(message, "userId") ||
+        get(message, "context.traits.userId") ||
+        get(message, "context.traits.id") ||
+        get(message, "anonymousId");
+      window.pdst("alias", {
+        id: externalId,
```
Comment (shrouti1507) [inline integrations/Podsights/browser.js:58]: let's hash it with sha256, as podsights support told us to choose our own hash algorithm. Let's mention that doc

Comment (sauravlal15) [reply integrations/Podsights/browser.js]: Done

File: integrations/Podsights/browser.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+  /**
+   * Page.
+   * for supporting path of Page
+   * @param {Page} page
+   */
+  page(rudderElement) {
+    const { properties } = rudderElement.message;
+    logger.debug("===In Podsights Page===");
+    window.pdst("view", properties);
```
[the remark this answers is not in the document]
Comment (sauravlal15) [reply integrations/Podsights/browser.js]: added

File: integrations/Podsights/browser.js [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+      get(message, "userId") ||
+      get(message, "context.traits.userId") ||
+      get(message, "context.traits.id") ||
+      get(message, "anonymousId");
+
+    events.forEach((podsightEvent) => {
+      if (podsightEvent === "lead") {
+        window.pdst(podsightEvent, { category: trimmedEvent, ...properties });
+      } else {
+        window.pdst(podsightEvent, properties);
```
Comment (shrouti1507) [inline integrations/Podsights/browser.js:78]: we are sending free flowing field from properties, but according to their documentation [here](https://podsights.com/docs/), they have some specific format, why we are not maintaining that?

Comment (sauravlal15) [reply integrations/Podsights/browser.js]: Added mapping, earlier I thought, the user, will pass all fields from properties. 
And also, should we support other fields apart from mapping?

File: integrations/Podsights/constants.js [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    destKey: "currency",
+    sourceKeys: "currency",
+  },
+  {
+    destKey: "discount_code",
+    sourceKeys: "discountCode",
+  },
+  {
+    destKey: "line_items",
+    sourceKeys: "lineItems",
```
Comment (shrouti1507) [inline integrations/Podsights/constants.js:136]: Line items should contain product object if any. This is what we see from their documentation:
```
pdst('checkout', {
  value: 10.0, // price
  currency: 'USD',
  discount_code: 'PODCAST_CODE',

  // optional line_items, if you would like advanced analytics.
  quantity: 3,
  line_items: [
    {
      value: 21.2,
      quantity: 1,
      product_id: '{{product.id}}',
      product_name: '{{product.title}}',
      product_type: '{{product.type}}',
      product_vendor: '{{product.vendor}}',
      variant_id: '{{variant.id}}',
      variant_name: '{{variant.title}}',
    },
    {
      value: 42.4,
      quantity: 2,
      product_id: '{{product.id}}',
      product_name: '{{product.title}}',
      product_type: '{{product.type}}',
      product_vendor: '{{product.vendor}}',
      variant_id: '{{variant.id}}',
      variant_name: '{{variant.title}}',
    },
  ],});
```
check what happens with these events if line_items are not sent

File: integrations/Podsights/constants.js [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    destKey: "discount_code",
+    sourceKeys: "discountCode",
+  },
+  {
+    destKey: "quantity",
+    sourceKeys: "quantity",
+  },
+  {
+    destKey: "line_items",
+    sourceKeys: "lineItems",
```
Comment (shrouti1507) [inline integrations/Podsights/constants.js:117]: same as the purchase event. If product array is not there, lets search from message.properties itself

File: integrations/Podsights/constants.js [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    destKey: "type",
+    sourceKeys: "type",
+  },
+  {
+    destKey: "category",
+    sourceKeys: "category",
+  },
+];
+
+const PRODUCT_EVENT = [
```
Comment (shrouti1507) [inline integrations/Podsights/constants.js:32]: I hope we have mentioned in docs that for product events, we are expecting product properties to be present in message.properties?

File: integrations/Podsights/browser.js [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    const events = eventsMapping[trimmedEvent] || [];
+    if (events.length === 0) {
+      logger.error(`===No Podsights Pixel mapped event found. Aborting!===`);
+      return;
+    }
+    const externalId =
+      get(message, "userId") ||
+      get(message, "context.traits.userId") ||
+      get(message, "context.traits.id") ||
+      get(message, "anonymousId");
```
Comment (shrouti1507) [inline integrations/Podsights/browser.js:80]: Let's not use `anonymousId` if we do not have confirmation that anonymous users can merged after the user gets identified. Let's cross check this

File: integrations/Podsights/browser.js (hunk 3 above)
Comment (shrouti1507) [reply integrations/Podsights/browser.js]: > And also, should we support other fields apart from mapping
we need to confirm this from support team, if fields other than what mentioned in the doc is recorded at all. If yes, we can send.

File: integrations/Podsights/constants.js (hunk 4 above)
Comment (sauravlal15) [reply integrations/Podsights/constants.js:136]: `line_items` is not visible in the debugger section for any event, and it doesn't cause any effect if we pass it or not.

File: integrations/Podsights/browser.js (hunk 7 above)
Comment (sauravlal15) [reply integrations/Podsights/browser.js:80]: Okay, will remove `anonymousId`.

File: integrations/Podsights/browser.js (hunk 3 above)
Comment (sauravlal15) [reply integrations/Podsights/browser.js]: As in the network call, I do not see any other attributes which they accept, and also in the dashboard, some limited fields are visible that we already mapped. So should we ask them? @shrouti1507

File: integrations/Podsights/constants.js (hunk 6 above)
Comment (sauravlal15) [reply integrations/Podsights/constants.js:32]: Should we consider this case:
```
product: [
 {
     product_name: "",
     product_type: "",
     product_vendor: "", 
     product_id: "",
 },
 {
     product_name: "",
     product_type: "",
     product_vendor: "", 
     product_id: "",
 } 
]
```
And from this, should we'll fetch each item detail and call the product_event same as the product object size?
@shrouti1507

File: integrations/Podsights/constants.js (hunk 5 above)
Comment (sauravlal15) [reply integrations/Podsights/constants.js:117]: This event doesn't take any product properties, so we are not passing any product array.

File: integrations/Podsights/constants.js (hunk 4 above)
Comment (sauravlal15) [reply integrations/Podsights/constants.js:136]: As mentioned in the docs, should we take `line_items` object values from `message.properties`?

File: integrations/Podsights/constants.js (hunk 5 above)
Comment (sauravlal15) [reply integrations/Podsights/constants.js:117]: Same for this also, `line_items` doesn't have any effects in this event, So, As mentioned in the docs, should we take `line_items` object values from message.properties?

File: integrations/Podsights/browser.js (hunk 3 above)
Comment (shrouti1507) [reply integrations/Podsights/browser.js]: you mean, the attributes other than whatever they mentioned in the docs?

File: integrations/Podsights/constants.js (hunk 5 above)
Comment (shrouti1507) [reply integrations/Podsights/constants.js:117]: do you see line_items not going in the network call?

File: integrations/Podsights/constants.js (hunk 4 above)
Comment (shrouti1507) [reply integrations/Podsights/constants.js:136]: if product array is present, then from message.properties.products if not then message.properties

File: integrations/Podsights/constants.js (hunk 6 above)
Comment (shrouti1507) [reply integrations/Podsights/constants.js:32]: from each product object, we should fetch only these keys and we have to mention the mapping in the doc.

File: integrations/Podsights/constants.js (hunk 5 above)
Comment (shrouti1507) [reply integrations/Podsights/constants.js:117]: Just verified, line_items are going, not showing in the dashboard, but going on network calls. Also any other field, except what has been listed in the docs, are not going. So let us stick to the exact fields that they have said

File: integrations/Podsights/constants.js (hunk 6 above)
Comment (shrouti1507) [reply integrations/Podsights/constants.js:32]:
```product: [
 {
     product_name: "",
     product_type: "",
     product_vendor: "", 
     product_id: "",
 },
 {
     product_name: "",
     product_type: "",
     product_vendor: "", 
     product_id: "",
 } 
]
```
this case need not to be considered. If this case arrives, we will omit the products array

File: integrations/Podsights/browser.js (hunk 3 above)
Comment (sauravlal15) [reply integrations/Podsights/browser.js]: Yes

File: integrations/Podsights/constants.js (hunk 4 above)
Comment (sauravlal15) [reply integrations/Podsights/constants.js:136]: Okay, got it, will update

File: integrations/Podsights/constants.js (hunk 5 above)
Comment (sauravlal15) [reply integrations/Podsights/constants.js:117]: Okay, will update

File: integrations/Podsights/constants.js (hunk 6 above)
Comment (sauravlal15) [reply integrations/Podsights/constants.js:32]: done

File: integrations/Podsights/browser.js [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+      get(message, "userId") ||
+      get(message, "context.traits.userId") ||
+      get(message, "context.traits.id");
+    let payload;
+    events.forEach((podsightEvent) => {
+      switch (podsightEvent.trim().toLowerCase()) {
+        case "lead":
+          payload = constructPayload(properties, LEAD_EVENT);
+          window.pdst(podsightEvent, payload);
+          this.loadAliasEvent(externalId);
```
Comment (shrouti1507) [inline integrations/Podsights/browser.js:95]: don't we need to check if the user has enabled it or not?

File: integrations/Podsights/browser.js [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+          break;
+        case "purchase": {
+          payload = payloadBuilder(properties, PURCHASE_EVENT);
+          window.pdst(podsightEvent, payload);
+          this.loadAliasEvent(externalId);
```
Comment (shrouti1507) [inline integrations/Podsights/browser.js:100]: here also

File: integrations/Podsights/browser.js [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    let payload;
+    events.forEach((podsightEvent) => {
+      switch (podsightEvent.trim().toLowerCase()) {
+        case "lead":
+          payload = constructPayload(properties, LEAD_EVENT);
+          window.pdst(podsightEvent, payload);
+          this.loadAliasEvent(externalId);
+          break;
+        case "purchase": {
+          payload = payloadBuilder(properties, PURCHASE_EVENT);
```
Comment (shrouti1507) [inline integrations/Podsights/browser.js:98]: have we checked what happens if we pass line_items in purchase events as well? does it go via the network call?

File: integrations/Podsights/browser.js [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+  track(rudderElement) {
+    logger.debug("===In Podsights Track===");
+    const { message } = rudderElement;
+    const { event, properties } = message;
+    if (!event) {
+      logger.error("[Podsights]: event name from track call is missing.");
+      return;
+    }
+
+    const eventsMapping = getHashFromArrayWithDuplicate(
```
Comment (shrouti1507) [inline integrations/Podsights/browser.js:73]: let us also add some default mapping. The priority will be given to user defined mapping, and if not found there we can fallback to our default mapping. if that is also not matching then throw error

File: integrations/Podsights/browser.js (hunk 8 above)
Comment (sauravlal15) [reply integrations/Podsights/browser.js:95]: It was checking inside the function.

File: integrations/Podsights/browser.js (hunk 9 above)
Comment (sauravlal15) [reply integrations/Podsights/browser.js:100]: same

File: integrations/Podsights/browser.js (hunk 11 above)
Comment (sauravlal15) [reply integrations/Podsights/browser.js:73]: Does default mapping mean with RudderStack standard event?

File: integrations/Podsights/browser.js (hunk 10 above)
Comment (sauravlal15) [reply integrations/Podsights/browser.js:98]: Yes, it's going via network call.

File: integrations/Podsights/browser.js (hunk 11 above)
Comment (sauravlal15) [reply integrations/Podsights/browser.js]: done

**Review by shrouti1507 — APPROVED** [review]

File: integrations/Podsights/browser.js (hunk 11 above)
Comment (shrouti1507) [reply integrations/Podsights/browser.js]: lets add this in research spec and let the doc team know

**Review by MoumitaM — APPROVED** [review]

**Pull request merged**